File: Dynamic-Programming/322. Coin Change/coinChange.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function (coins, amount) {
    let dp = new Array(amount + 1).fill(amount + 1);
    let n = dp.length;
    dp[0] = 0;

    for (let i = 1; i < n; i++) {
        for (const coin of coins) {
            if (i - coin >= 0) {
                dp[i] = Math.min(dp[i], 1 + dp[i - coin]);
            }
        }
    }

    return dp[n - 1] === amount + 1 ? -1 : dp[n - 1];
};

File: Structures & Algo/D. Dynamic Programming/322. Coin Change/coinChange.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        memo = [-666 for _ in range(amount + 1)]
        
        def dp(coins, amount):
            if amount == 0: return 0 
            if amount < 0: return -1
            
            if memo[amount] != -666:
                return memo[amount]
            
            res = math.inf
            for coin in coins:
                sub = dp(coins, amount - coin)
                if sub == -1: continue
                res = min(res, sub + 1)
            
            memo[amount] = res if res != math.inf else -1
            return memo[amount]
        
        return dp(coins, amount)

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount + 1 for _ in range(amount + 1)]
        dp[0] = 0
        
        for i in range(1, len(dp)):
            for coin in coins:
                if i - coin < 0: continue
                dp[i] = min(dp[i], dp[i - coin] + 1)
        
        return dp[-1] if dp[-1] != amount + 1 else -1
            

